### PR TITLE
Feature/bk core tag functionality

### PR DESCRIPTION
**Pull request by burkaygur**
NEW:

Updated README and CHANGELOG. New "Initialize SDK" section. And more cleanups.

The new iOS SDK:

1) Introduced a new initializer: initDirectAutoIdfaEnabledWithSiteId that enables direct connection to the tags endpoint. HTTPS is enabled by default.

2) Enabled non-ARC (manual garbage collection). In order to use this mode, add the compiler flag to Bluekai.m -fno-objc-arc

3) Fixed memory leaks around SBJson (later on removed completely and replaced with native code) and Reachability third party libraries.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,29 +71,55 @@ In `ViewController.h` file, define an instance of BlueKai SDK.
...
 blueKaiSdk = [[BlueKai alloc] initWithSiteId:@"2" withAppVersion:version withView:self withDevMode:YES]; 
 ```
+or 
+```objective-c
+blueKaiSDK = [[BlueKai alloc] initWithSiteId:@"2" withAppVersion:version withIdfa:idfa withView:self withDevMode:YES];
+```
+if you would like to provide the Apple IDFA.
 
-The first argument (`initWithSiteId`) is site id, which you would get from BlueKai.
+The argument (`initWithSiteId`) is site id, which you would get from BlueKai.
```
Comment (antonbk) [inline README.md]: The `initWithSiteId` argument is site id, which you would get from BlueKai.

File: README.md [hunk 2]
```diff
@@ -71,29 +71,55 @@ In `ViewController.h` file, define an instance of BlueKai SDK.
...
 
-The second argument is app version (`withAppVersion`) and is not necessarily the
+The argument is app version (`withAppVersion`) and is not necessarily the
```
Comment (antonbk) [inline README.md]: The `withAppVersion` argument is app version and is not necessarily the

File: README.md [hunk 3]
```diff
@@ -71,29 +71,55 @@ In `ViewController.h` file, define an instance of BlueKai SDK.
...
 application version of the calling application. This is a value by
 which BlueKai can uniquely indentify the application from which the
 request originated. A suggested approach is to use "app name-version_number" format.
 
-The third argument (`withView`) is a view to which the SDK can attach an invisible WebView to call BlueKai's tag. When
+The argument (`withView`) is a view to which the SDK can attach an invisible WebView to call BlueKai's tag. When
```
Comment (antonbk) [inline README.md]: Same here - needs better wording.

File: README.md [hunk 4]
```diff
@@ -71,29 +71,55 @@ In `ViewController.h` file, define an instance of BlueKai SDK.
...
 `devMode` is enabled, this view becomes visible to display values being passed to BlueKai's server for debugging.
 
-The last argument (`withDevMode`) indicates whether you want developer mode. In this mode, a webview overlay will be displayed 
+**[OPTIONAL]** The argument (`withIdfa`) is the Apple IDFA of the device, which provides a way of unique identification. The host app must provide this value.
```
Comment (antonbk) [inline README.md]: `withIdfa ` argument ...

File: README.md [hunk 5]
```diff
@@ -129,29 +155,29 @@ Define a method appCameToForeground and call `resume()`:
...
 
 Declare the BlueKai SDK delegate in `ViewController.h`. This step is
-optional and is needed only if you need a notification when data is posted
+optional and is needed only if you need a callback when data is posted
 to BlueKai server.
 
 
-```objective-c
-@protocol BlueKaiOnDataPostedListener;
+```objecidfative-c
```
Comment (antonbk) [inline README.md]: objective-c ?

File: README.md [hunk 6]
```diff
@@ -71,29 +71,55 @@ In `ViewController.h` file, define an instance of BlueKai SDK.
...
+The argument (`withView`) is a view to which the SDK can attach an invisible WebView to call BlueKai's tag. When
 `devMode` is enabled, this view becomes visible to display values being passed to BlueKai's server for debugging.
 
-The last argument (`withDevMode`) indicates whether you want developer mode. In this mode, a webview overlay will be displayed 
+**[OPTIONAL]** The argument (`withIdfa`) is the Apple IDFA of the device, which provides a way of unique identification. The host app must provide this value.
+
+The argument (`withDevMode`) indicates whether you want developer mode. In this mode, a webview overlay will be displayed 
 with response from the BluaKai server. You should turn this feature off in your production code.
 
+#### **NEW!** 2. NSURLConnection - Direct Connection
```
Comment (antonbk) [inline README.md]: No need to have `**NEW!**`. It's more suitable for change long.
Readme should have generic information about available features.

File: README.md [hunk 7]
```diff
@@ -210,80 +236,88 @@ Use HTTPS transfer protocol
 
 ### Methods
 
+**NEW!** Create the instance for Bluekai SDK with required arguments for direct connection (IDFA is grabbed automatically if user has not disabled it). This method is recommended for applications that require high memory efficiency. Uses NSURLConnection vs. UIWebView.
```
Comment (antonbk) [inline README.md]: ~~NEW!~~

File: sample/BlueKai.h [hunk 8]
```diff
@@ -21,146 +21,215 @@
...
+*/
+- (id)initWithSiteId:(NSString *)siteID
+      withAppVersion:(NSString *)version
+            withIdfa:(NSString *)idfa
+            withView:(UIViewController *)view
+         withDevMode:(BOOL)devMode;
+
+/** Init BlueKai SDK (Deprecated)
+*
+* Deprecated; Use "initWithSiteId" instead
```
Comment (antonbk) [inline sample/BlueKai.h]: There is a way to mark methods as deprecated:
see http://stackoverflow.com/questions/3908715/how-do-i-flag-a-method-as-deprecated-in-objective-c-2-0#answer-3908766

File: sdk/Bluekai_SDK/BlueKai.m [hunk 9]
```diff
@@ -2,13 +2,13 @@
...
 #import "BlueKai_Reachability.h"
-#import "BlueKai_SBJSON.h"
+#import <AdSupport/AdSupport.h>
 
 @implementation BlueKai {
     BOOL _alertShowBool,
          _webLoaded;
 
-    int _urlLength,
+    unsigned long _urlLength,
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: overkill, URLs cannot be that long

File: sdk/Bluekai_SDK/BlueKai.m [hunk 10]
```diff
@@ -100,11 +85,10 @@ - (id)initWithSiteId:(NSString *)siteID withAppVersion:(NSString *)version withI
             [[NSFileManager defaultManager] createFileAtPath:atmt_fileAtPath contents:nil attributes:nil];
         }
 
-        _keyValDict = [[NSMutableDictionary alloc] initWithDictionary:[self getKeyValueDictionary:[self readStringFromKeyValueFile]]];
+        _keyValDict = [[NSMutableDictionary alloc] initWithDictionary:[self jsonStringToDictionary:[self readStringFromFile:@"user_data.bk"]]];
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: Make `user_data.bk` a constant.

File: sdk/Bluekai_SDK/BlueKai.m [hunk 11]
```diff
@@ -126,6 +110,109 @@ - (id)initWithSiteId:(NSString *)siteID withAppVersion:(NSString *)version withV
...
+        
+        if (![_userDefaults objectForKey:@"userIsOptIn"]) {
+            [_userDefaults setObject:(_optInPreference ? @"YES" : @"NO") forKey:@"userIsOptIn"];
+        }
+        
+        [self saveSettings:nil];
+        
+        // check the dictionary for previous values
+        NSString *filePath = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES)[0];
+        NSString *fileName = @"user_data.bk";
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: Use constant for string `user_data.bk`

File: sdk/Bluekai_SDK/BlueKai.m [hunk 12]
```diff
@@ -126,6 +110,109 @@ - (id)initWithSiteId:(NSString *)siteID withAppVersion:(NSString *)version withV
...
+         withDevMode:(BOOL)devMode{
+    return [self initDirectWithSiteId:siteID withAppVersion:version withIdfa:idfa withUserAgent:NULL withDevMode:devMode];
+}
+
+- (id)initDirectAutoIdfaEnabledWithSiteId:(NSString *)siteID withAppVersion:(NSString *)version withDevMode:(BOOL)devMode{
+    //TO-DO Add more properties about User-Agent here
+    NSString *userAgent = @"iOS BlueKaiSDK";
+    //UIWebView* webView = [[UIWebView alloc] initWithFrame: CGRectMake(0,0,0,0)];
+    //NSString* secretAgent = [webView stringByEvaluatingJavaScriptFromString:@"navigator.userAgent"];
+    //[webView release];
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: remove comments

File: sdk/Bluekai_SDK/BlueKai.m [hunk 13]
```diff
@@ -126,6 +110,109 @@ - (id)initWithSiteId:(NSString *)siteID withAppVersion:(NSString *)version withV
...
+    _useHttps = YES;
+    return [self initDirectWithSiteId:siteID withAppVersion:version withIdfa:[self identifierForAdvertising] withUserAgent:userAgent withDevMode:devMode];
+}
+
+// credit: http://stackoverflow.com/questions/12944504/
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: here is official doc reference: https://developer.apple.com/library/ios/documentation/AdSupport/Reference/ASIdentifierManager_Ref/

File: sdk/Bluekai_SDK/BlueKai.m [hunk 14]
```diff
@@ -166,7 +240,7 @@ - (void)resume {
         _webUrl = [[NSMutableString alloc] init];
     }
 
-    _keyValDict = [[NSMutableDictionary alloc] initWithDictionary:[self getKeyValueDictionary:[self readStringFromKeyValueFile]]];
+    _keyValDict = [[NSMutableDictionary alloc] initWithDictionary:[self jsonStringToDictionary:[self readStringFromFile:@"user_data.bk"]]];
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: Use constant for `user_data.bk`

File: sdk/Bluekai_SDK/BlueKai.m [hunk 15]
```diff
@@ -878,32 +973,61 @@ - (void)uploadIfNetworkIsAvailable {
...
+        
+        #if !__has_feature(objc_arc)
+        urlString = [[[NSMutableString alloc] initWithString:[NSString stringWithFormat:@"%@%@%@?site=%@&", protocol, serverURL, endPoint, _siteId]] autorelease];
+        #else
+        urlString = [[NSMutableString alloc] initWithString:[NSString stringWithFormat:@"%@%@%@?site=%@&", protocol, serverURL, endPoint, _siteId]];
+        #endif
+    }
 
-    [urlString appendString:[NSString stringWithFormat:@"appVersion=%@", _appVersion]];
+    [urlString appendString:[self getDataParam:@"phint" WithKey:@"appVersion" AndValue:_appVersion]];
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: Is it confirmed `appVersion` to be sent as a phint? Usually pixel servers have special treatment for non-phint query params, whereas phints are treated with classification rules and tag with taxonomy categories.

File: sdk/Bluekai_SDK/BlueKai.m [hunk 16]
```diff
@@ -918,4 +1042,81 @@ - (void)blueKaiLogger:(BOOL)devMode withString:(NSString *)string withObject:(NS
...
+    if(_useDirectHTTPCalls) {
+        if(value != NULL) {
+            if (type != NULL){
+                return [NSString stringWithFormat:@"%@=%@", type, [self urlEncode:[NSString stringWithFormat:@"%@=%@", key, value]]];
+            }
+            else{
+                return [NSString stringWithFormat:@"%@=%@", [self urlEncode:key], [self urlEncode:value]];
+            }
+        } else {
+            return [NSString stringWithFormat:@"%@=%@", type, key];
```
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: Should we encode key?
```objective-c
[self urlEncode:key]
```

File: sdk/Bluekai_SDK/BlueKai.m (hunk 15 above)
Comment (antonbk) [inline sdk/Bluekai_SDK/BlueKai.m]: If it's a phint, than use `getBlueKaiParamWithKey `

File: sdk/Bluekai_SDKTests/BlueKaiTestCase.m [hunk 17]
```diff
@@ -228,7 +227,7 @@ - (void)testCanReadStringFromKeyValueFile
     NSString *fileAtPath = [filePath stringByAppendingPathComponent:fileName];
     [[expected dataUsingEncoding:NSUTF8StringEncoding] writeToFile:fileAtPath atomically:NO];
 
-    XCTAssertEqualObjects(expected, [blueKaiSdk readStringFromKeyValueFile], @"Data read from file differ from expected string");
+    XCTAssertEqualObjects(expected, [blueKaiSdk readStringFromFile:@"user_data.bk"], @"Data read from file differ from expected string");
```
Comment (antonbk) [inline sdk/Bluekai_SDKTests/BlueKaiTestCase.m]: Re-use constant for `"user_data.bk"`

**Comment by antonbk** [comment]
Just a minor things, mostly refactoring to remove some duplication, than :shipit:

File: sdk/Bluekai_SDK/BlueKai.m (hunk 16 above)
Comment (burkaygur) [inline sdk/Bluekai_SDK/BlueKai.m]: I decided to encode that, because there could be spaces or some other special characters.

**Pull request closed without merging**